tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/GraphQLSecurityCommon.cs

```diff
@@ -0,0 +1,117 @@
+using Datadog.Trace.AppSec;
+using Datadog.Trace.AppSec.Coordinator;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.GraphQL;
+
+internal sealed class GraphQLSecurityCommon
+{
+    private static readonly Lazy<GraphQLSecurityCommon> LazyInstance = new(() => new GraphQLSecurityCommon());
+
+    private readonly ConcurrentDictionary<IScope, Dictionary<string, object>> _scopeResolvers;
+
+    private GraphQLSecurityCommon()
+    {
+        _scopeResolvers = new ConcurrentDictionary<IScope, Dictionary<string, object>>();
+    }
+
+    private static GraphQLSecurityCommon Instance
+        => LazyInstance.Value;
+
+    private static Dictionary<string, object> PopScope(IScope scope)
+    {
+        var resolvers = Instance.GetScopeResolvers(scope);
+        Instance.RemoveScopeResolvers(scope);
+        return resolvers;
+    }
+
+    private Dictionary<string, object> GetScopeResolvers(IScope scope)
+    {
+        if (!_scopeResolvers.TryGetValue(scope, out var resolvers))
+        {
+            resolvers = new Dictionary<string, object>();
+            _scopeResolvers.TryAdd(scope, resolvers);
+        }
+
+        return resolvers;
+    }
+
+    private void RemoveScopeResolvers(IScope scope)
+    {
+        _scopeResolvers.TryRemove(scope, out _);
+    }
+
+    /// <summary>
+    /// Register a called resolver with its arguments for a given scope into the <see cref="_scopeResolvers"/> dictionary.
+    /// </summary>
+    internal static void RegisterResolverCall(IScope scope, string resolverName, Dictionary<string, object> arguments)
+    {
+        var resolvers = Instance.GetScopeResolvers(scope);
+
+        if (!resolvers.TryGetValue(resolverName, out var resolverCalls))
+        {
+            resolverCalls = new List<object> { arguments };
+            resolvers.Add(resolverName, resolverCalls);
+        }
+        else
+        {
+            // Add the current resolver call with arguments
+            ((List<object>)resolverCalls).Add(arguments);
+        }
+    }
+
+    /// <summary>
+    /// Run the WAF for the given scope of the GraphQL request.
+    /// </summary>
+    public static void RunSecurity(Scope scope)
+    {
+        if (!IsEnabled())
+        {
+            return;
+        }
+
+        var security = Security.Instance;
+        var allResolvers = PopScope(scope);
+
+        var args = new Dictionary<string, object> { { "graphql.server.all_resolvers", allResolvers } };
+#if NETFRAMEWORK
+        var httpContext = HttpContext.Current;
+#else
+        var httpContext = CoreHttpContextStore.Instance.Get();
+#endif
+        var securityCoordinator = new SecurityCoordinator(security, httpContext, scope.Span);
+        securityCoordinator.Check(args);
+    }
+
+    /// <summary>
+    /// Check if we need to perform an analysis for the GraphQL request.
+    /// </summary>
+    /// <returns>True if ASM is enabled and the current request is not a WebSocket connection.</returns>
+    public static bool IsEnabled()
```

Comment:
Feels like this is in the wrong place, as it's nothing to do with the scope handling that the rest of this class deals with

Reply:
I agree is should be split out, but can't really find any better place for the moment, I'll bear this in mind for future changes to GraphQL. We should be add more functionality soon.